test/partition-healing-tests.js

```diff
@@ -0,0 +1,357 @@
+                dsl.waitForPingResponse(t, tc, 1),
+                dsl.waitForPingResponse(t, tc, 2),
+                dsl.assertStats(t, tc, 2, 0, 2),
+
+                // According to B, a1 is faulty.
+                makeAFaultyInB(t, tc),
+
+                // Trigger partition heal
+                dsl.callEndpoint(t, tc, "/admin/healpartition/disco", {},
+                        function(eventBody) { validateHealRequest(t, tc, eventBody) }),
+                waitForHealPartitionDiscoResponse(t, tc),
+
+                // Verify join request came in to b1 or b2
+                dsl.waitForJoins(t, tc, 1),
+
+                // Verify b1 and b2 are declared suspects, but 'a' is not disseminated as alive.
+                verifySuspects(t, tc, [1, 2], [0]),
+
+                // SUT tries to reincarnate partition A -- send a ping to a1.
+                // Also verify it didn't have information about b1: that means
+                // sut didn't try to merge partitions yet.
+                verifySuspects(t, tc, [0], [1, 2])
+            ]
+        }
+        )
+     );
+
+// Assign names to nodes and update membership of B from test coordinator.
+function makeAFaultyInB(t, tc) {
+    return function makeAFaultyInB(list, cb) {
+        tc.fakeNodes[1].status = 'alive';
+        tc.fakeNodes[2].status = 'alive';
+
+        var B = tc.getMembership();
+        // According to B, a1 is faulty
+        _.find(B, {port: tc.fakeNodes[0].port}).status = 'faulty';
+
+        tc.fakeNodes[1].membership = _.cloneDeep(B);
+        tc.fakeNodes[2].membership = _.cloneDeep(B);
+
+        cb(list);
+    }
+}
+
+function waitForHealPartitionDiscoResponse(t, tc) {
+    return function waitForHealPartitionDiscoResponse(list, cb) {
+        var d = _.find(list, { type: events.Types.AdminHealPartitionDisco });
+        if (d === undefined) {
+            cb(null);
+        } else {
+            list.splice(list.indexOf(d), 1);
+            cb(list);
+        }
+    }
+}
+
+// verifySuspects waits for a ping from sut that contains the following changes:
+// - all of 'nodes' are declared suspect, and
+// - no nodes from 'nodes_inother_partition' are declared alive
+function verifySuspects(t, tc, nodes, nodes_in_other_partition) {
+    return function verifyBsuspects(list, cb) {
+        // Which addresses should be declared suspects in the same message.
+        var addresses = _.map(nodes, function(ix) {
+            return tc.fakeNodes[ix].getHostPort();
+        });
+
+        var addresses_in_other_partition = _.map(nodes, function(ix) {
+            return tc.fakeNodes[ix].getHostPort();
+        });
+
+        var pingRequests = _.filter(list, {
+            type: events.Types.Ping,
+            direction: 'request'
+        });
+
+        var found_all_suspects = false;
+        _.forEach(pingRequests, function(pingRequest) {
+            var suspects = [];
+            _.forEach(safeJSONParse(pingRequest.arg3).changes, function(change) {
```

Comment:
instead of safeparsing you can use `pingRequest.body`. It is already safe parsed in the `events.js`